cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/PipelinePlanner.java

```diff
@@ -214,203 +206,312 @@ public PipelinePlan plan(PipelineSpec spec) {
+              continue;
+            }
+
+            if (controlNodes.containsAll(dag2.getNodes())) {
+              phaseConnections.add(new Connection(controlNodeName, dag2.getSinks().iterator().next(),
+                                                  condition));
+            } else {
+              phaseConnections.add(new Connection(controlNodeName, dag2Name, condition));
+            }
+            break;
           }
         }
       }
     }
 
-    Map<String, String> conditionConnectors = getConnectorsAssociatedWithConditions(conditionBranches.keySet(),
-                                                                                    conditionChildToParent);
+    Map<String, String> controlConnectors = getConnectorsAssociatedWithConditions(conditionBranches.keySet(),
+                                                                                  conditionChildToParent,
+                                                                                  conditionInputs, conditionOutputs,
+                                                                                  actionNodes);
+
+    // At this point we have added phase connections between subdags, processing control nodes if they are
+    // common to the two dags as either source or sink. However subdag can only contain control nodes
+    // when two contol nodes are chained together for example Action-->Condition. We need to add connection
+    // between them. Also we need to process any control nodes which are NOT yet in the processedControlNodes set.
+    // For example if we have two splitted dags
+    // dag1: Action-->Condition
+    // dag2: Condition->N0
+    // Now we processed connections from Condition, since its common between dag 1 sink and dag 2 source.
+    // However we have not process Action yet, since there is not dag0 with which it is common. So the Action
+    // will not be in processedControlNode set.
     for (Map.Entry<String, Dag> dagEntry : subdags.entrySet()) {
-      // If dag only contains conditions then ignore it
-      if (conditionBranches.keySet().containsAll(dagEntry.getValue().getNodes())) {
+      Dag dag = dagEntry.getValue();
+      if (controlNodes.containsAll(dag.getSources())) {
+        String source = dag.getSources().iterator().next();
+        String sink = dagEntry.getKey();
+        if (controlNodes.containsAll(dag.getNodes())) {
+          // Current dag only contain control nodes.
+          sink = dag.getSinks().iterator().next();
+        }
+        if (!processedControlNodes.contains(source)) {
+          // This control node is not processed yet, because it is first in the pipeline.
+          ConditionBranches branches = conditionBranches.get(source);
+          if (branches != null) {
+            // this is condition node
+            boolean condition = false;
+            if (branches.getTrueOutput() != null && dag.getNodes().contains(branches.getTrueOutput())) {
```

Comment:
can be simplified to:
```
boolean condition = branches.getTrueOutput() != null && dag.getNodes().contains(branches.getTrueOutput());
```
also, why do you need to check that it's not null? does dag.getNodes() not allow nulls in it?